Document/0x05b-Basic-Security_Testing.md

```diff
@@ -345,6 +345,76 @@ Successfully installed 1 modules, 0 already installed
+
+##### Preparation
+
+For a full dynamic analysis of an Android app also FCM should be intercepted. To be able to intercept the messages several steps should be considered for preparation. The following example can be used on Mac OS X:
+
+* Install the CA certificate of your interception proxy into your Android phone<sup>[2]</sup>.
+* A Man-in-the-middle attack should be executed so all traffic from the mobile device is redirected to your testing machine. This can be done by using a tool like ettercap<sup>[24]</sup>. It can be installed by using brew.
+
+```bash
+$ brew install ettercap
+```
+
+FCM can use two different protocols to communicate with the Google backend, either XMPP or HTTP.
+
+**HTTP**
+
+The ports used by FCM for HTTP are 5228, 5229, and 5230. Typically only 5228 is used, but sometimes also 5229 and 5230 is used.
+
+* Configure a local port forwarding on your machine for the ports used by FCM. The following example can be used on Mac OS X<sup>[23]</sup>:
+
+```bash
+$ echo "
+rdr pass inet proto tcp from any to any port 5228-> 127.0.0.1 port 8080
+rdr pass inet proto tcp from any to any port 5229 -> 127.0.0.1 port 8080
+rdr pass inet proto tcp from any to any port 5239 -> 127.0.0.1 port 8080
+" | sudo pfctl -ef -
+```
+
+* The interception proxy need to listen to the port specified in the port forwarding rule above, which is 8080.
+
+**XMPP**
+
+The ports used by FCM over XMPP are 5235 (Production) and 5236 (Testing)<sup>[26]</sup>.
+
+* Configure a local port forwarding on your machine for the ports used by FCM. The following example can be used on Mac OS X<sup>[23]</sup>:
+
+```bash
+$ echo "
+rdr pass inet proto tcp from any to any port 5235-> 127.0.0.1 port 8080
+rdr pass inet proto tcp from any to any port 5236 -> 127.0.0.1 port 8080
+" | sudo pfctl -ef -
+```
+
+* The interception proxy need to listen to the port specified in the port forwarding rule above, which is 8080.
+
+##### Intercepting Messages
+
+Your testing machine and the Android device need to be in the same wireless network. Start ettercap with the following command and replace the IP addresses with the one of the Android device and the network gateway in the wireless network.
+
+```bash
+$ ettercap -T -i eth0 -M arp:remote /192.168.0.1// /192.168.0.105//
+```
+
+Start using the app and trigger a function that uses FCM. You should see HTTP messages showing up in your interception proxy.
+
+![Intercepted Messages](Images/Chapters/0x05b/FCM_Intercept.png)
+
+If XMPP is used, Burp will not show this traffic, but either a Burp Plugin like Burp-non-HTTP-Extension<sup>[28]<sup> can be used or Mitm-relay<sup>[27]<sup> that leverages Burp to visualize XMPP traffic.
```

Comment:
Interception proxies will already be described in the "Mobile App Security Testing" chapter. Therefore I assume that the reader will already heard about it when he reaches this section :-)
But I rephrased the sentence to also include ZAP.